### PR TITLE
Adding support for private registry auth on stack deploy

### LINKED ISSUE #1425: Private registry authentication
`docker service create` has an option to pass local credentials to the swarm for image pulls (`--with-registry-auth`). We need a similar mechanism for `amp stack deploy`

### PR DESCRIPTION
**Pull request by bquenin**
Fixes #1425 

This PR enables providing registry authentication credentials on stack deploy by adding a `--with-registry-auth` option. Since credentials are carried over to `amplifier`, it would be better to add TLS support ASAP.

## How to test
Save the following stack file to `private.yml`:
```
version: "3"

networks:
  default:
    external:
      name: ampnet

services:

  pinger:
    image: appcelerator/private-pinger:${TAG:-latest}
    networks:
      default:
        aliases:
          - pinger
    environment:
      SERVICE_PORTS: "3000"
      VIRTUAL_HOST: "pinger.examples.*,https://pinger.examples.*"
    deploy:
      replicas: 3
      restart_policy:
        condition: on-failure
      placement:
        constraints: [node.labels.amp.type.user == true]
```

Then
```
$ ampmake build
$ amp cluster create -t local
$ docker logout
$ amp login --name su --password password
$ amp stack deploy -c private.yml
$ amp service ls
[user su @ localhost:50101]
ID                          NAME             MODE         REPLICAS   FAILED TASKS   STATUS   IMAGE                         TAG
yliw8vo3x9j2vkeia0c01q1su   private_pinger   replicated   0/3        6              ERROR    appcelerator/private-pinger   latest
$ amp stack rm private
$ docker login
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username: bquenin
Password: 
Login Succeeded
$ amp stack deploy -c private.yml
[user su @ localhost:50101]
Deploying stack private using private.yml
Creating service private_pinger

$ amp service ls
[user su @ localhost:50101]
ID                          NAME             MODE         REPLICAS   FAILED TASKS   STATUS   IMAGE                         TAG
eoxfdms8r1miwgpwi6wttel84   private_pinger   replicated   0/3        6              ERROR    appcelerator/private-pinger   latest
$ amp stack rm private
$ amp stack deploy -c private.yml --with-registry-auth
[user su @ localhost:50101]
Deploying stack private using private.yml
Creating service private_pinger

$ amp service ls
[user su @ localhost:50101]
ID                          NAME             MODE         REPLICAS   FAILED TASKS   STATUS    IMAGE                         TAG
10o7aghmoq09crl0mn6574y0r   private_pinger   replicated   3/3        0              RUNNING   appcelerator/private-pinger   latest
```

### CODE REVIEW

**Review by subfuzion — COMMENTED** [review]
Curious ... why are you adding the docker client package instead of importing it?

**Comment by bquenin** [comment]
@subfuzion In order to add the `SetConfigFile` method to the docker CLI.

File: pkg/docker/docker.go [hunk 1]
```diff
@@ -196,18 +201,34 @@ func (d *Docker) ComposeIsAuthorized(compose *types2.Config) bool {
...
+			// Read client configuration file from reader
+			cf := configfile.ConfigFile{
+				AuthConfigs: make(map[string]types.AuthConfig),
+			}
+			if err := json.NewDecoder(bytes.NewReader(configFile)).Decode(&cf); err != nil {
+				return err
+			}
+
+			// Update the cli with the new configuration
+			cli.SetConfigFile(&cf)
```
Comment (subfuzion) [inline pkg/docker/docker.go]: Can you be really explicit about marking what is custom code and rationale in comments around our modifications to the package (whether here or other locations)?

Comment (bquenin) [reply pkg/docker/docker.go]: sure

Comment (subfuzion) [reply pkg/docker/docker.go]: Thx ... and just to be clear ... there is no way to accomplish this by using our own type that embeds the Docker type and augments the behavior? This feels like driving a 2-penny nail with sledgehammer...

Comment (bquenin) [reply pkg/docker/docker.go]: I agree. Unfortunately, after some tests, it doesn't seem possible to do this.

Comment (subfuzion) [reply pkg/docker/docker.go]: That's a real shame

**Review by subfuzion — APPROVED** [review]
👍 My only concern now is having a strategy to keep pkg/docker current, but we need this now, so thanks.

**Comment by subfuzion** [comment]
@bquenin I forgot to ask, but please make sure our docs call this out as well. Thanks.

**Pull request merged**